Add a way to generate submissions

PhilippMatthes opened the pull request:
Closes: #221 

The new management command `generate_submissions` can be used to create user submissions for demo purposes, which means that it creates `Solutions` and `Users`. The users are generated with random animal names, like in Google Docs. 

@martinmo does this meet your issue or should the command be advanced to create more object types? (e.g. `SolutionFiles`)

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+
+    def create_solution(self, user, task, amount, start_date, end_date):
+        # Randomly choose, if the solution succeeded
+        passed = bool(random.randint(0, 1))
```
`random.choice([True, False])`

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+
+User = get_user_model()
+
+file_dir = os.path.dirname(os.path.abspath(__file__))
```
For new code, please use `pathlib`.

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        silent = options["silent"]
+        if start_date > end_date:
+            raise ValueError("Start date should be before end date.")
+        if not Task.objects.all().exists():
```
Can be shortened to `Task.objects.exists()` (is equivalent otherwise 👍).

martinmo commented:
> @martinmo does this meet your issue or should the command be advanced to create more object types? (e.g. SolutionFiles)

No, your current proposal is ok :-)

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+    dateformat = "%d.%m.%Y"
+
+    def create_users(self, amount):
+        with open("{}/animals.txt".format(file_dir), "r") as f:
```
With pathlib you should use:
```python
with file_dir.joinpath("animals.txt").open(mode="r") as f:
```
Also note the explicit `mode` keyword argument.

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        with open("{}/animals.txt".format(file_dir), "r") as f:
+            usernames = [animal.strip().lower() for animal in f.readlines()]
+        chosen_names = random.sample(usernames, amount)
+        users = list()
```
Replace `list()` with `[]`

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        with atomic():
+            users = self.create_users(users_amount)
+            tasks = Task.objects.all()
+            all_solutions = list()
```
Replace `list()` with `[]`

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        )
+        parser.add_argument(
+            "--solutions_amount",
+            help="Amount of solutions to generate",
```
*number* of solutions (amount is for uncountable nouns)

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        )
+        parser.add_argument(
+            "--users_amount",
+            help="Amount of users to generate",
```
*number* of users (amount is for uncountable nouns)

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+    def create_solution(self, user, task, amount, start_date, end_date):
+        # Randomly choose, if the solution succeeded
+        passed = random.choice([True, False])
+        # Generate random submission date between start and end date
```
Remove that this comment, imho it's obvious what the code does

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+        solution = Solution.objects.create(
+            author=user, task=task, passed=passed,
+        )
+        # If auto_now_add is active, the submission date must be supplied
```
What you meant is:
"If auto_now_add is active, the submission date must be supplied *after* object creation." (with stress on after)

What I "read" initially:
"auto_now_add doesn't work"

Proposal:
"Because of auto_now_add, we can't set a custom submission_date through the constructor."

(Reading through https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.DateField, in retroperspective it probably would have been better if I had used `default=timezone.now()` for this field.)

PhilippMatthes replied:
> Proposal:
> "Because of auto_now_add, we can't set a custom submission_date through the constructor."

Your proposal is better and does not imply erroneous behaviour of the `auto_now_add` option.

> (Reading through https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.DateField, in retroperspective it probably would have been better if I had used `default=timezone.now()` for this field.)

Yes, I can't remember where, but a few weeks ago, I ran into other problems using the `auto_now_add` option, because it automatically inherits `editable=False`.

martinmo commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+    def handle(self, *args, **options):
+        start_date = make_aware(datetime.strptime(options["start_date"], self.dateformat))
+        end_date = make_aware(datetime.strptime(options["end_date"], self.dateformat))
+        solutions_amount = options["solutions_amount"]
```
Still some `amounts` lying around (ten instances, to be precise). `grep -i amount generate_submissions.py` ;-)

PhilippMatthes commented on inloop/solutions/management/commands/generate_submissions.py:
```diff
@@ -0,0 +1,117 @@
+                ))
+        return users
+
+    def create_solution(self, user, task, amount, start_date, end_date):
```
Found another issue: the `amount` parameter is not used.

PhilippMatthes commented:
The animals list was removed, because it is too heavy for our purpose. I replaced the user names by `"GeneratedUser{}".format(i)`

martinmo commented:
> One last remaining thing now is that running pipenv run ./manage.py generate_submissions -h fails to print the help text.

I double checked that my changes aren't responsible for this, of course 👼

PhilippMatthes commented:
> I still found some issues and fixed them right away in 6 individual commits (so you can easily see what I changed and why). Please have a look at them and squash them together into one when you are done.
> 
> One last remaining thing now is that running `pipenv run ./manage.py generate_submissions -h` fails to print the help text.

Ok, I read your changes. Will squash them once the last fix is committed. The reason, why the help text wouldn't render is, because the dateformat contains literals used for string formatting. I added a top level attribute `help_text_safe_dateformat` and now it works.

PhilippMatthes commented:
@martinmo you have to teach me, how to commit something with a co-author.
